File: lib/iriq/recognizer_proposal.rb
require "set"

module Iriq
  # A suggestion that a new Recognizer should be added to the system.
  #
  # Emitted by Corpus#propose_recognizers. NOT automatically activated —
  # proposals carry enough evidence for a human to judge whether to add
  # the Recognizer to the built-in set (or, later, to register it
  # dynamically via a public Recognizer registry).
  #
  #   prefix             — the detected shape signature (e.g. "ghp_")
  #   suggested_type     — Symbol name we'd register the Recognizer under
  #                        if accepted (e.g. :ghp)
  #   positions          — every Position where the proposal matched
  #   hosts              — distinct hosts the proposal was seen at; a high
  #                        count is strong evidence the pattern isn't
  #                        host-local
  #   coverage           — fraction of sampled observations at affected
  #                        Positions matching the proposal pattern
  #   observation_count  — total matching observations across positions
  #   sample_values      — up to 5 example matches, for the human reviewer
  #   strategy           — the ProposalStrategy that emitted this record
  class RecognizerProposal
    attr_reader :prefix, :suggested_type, :positions, :hosts,
                :coverage, :confidence, :observation_count,
                :sample_values, :strategy

    def initialize(prefix:, suggested_type:, positions:, hosts:,
                   coverage:, observation_count:, sample_values:,
                   strategy:, confidence: nil)
      @prefix            = prefix
      @suggested_type    = suggested_type
      @positions         = positions.freeze
      @hosts             = hosts.is_a?(Set) ? hosts.dup.freeze : Set.new(hosts).freeze
      @coverage          = coverage
      @observation_count = observation_count
      @sample_values     = sample_values.freeze
      @strategy          = strategy
      @confidence        = confidence.nil? ? compute_confidence : confidence
    end

    def to_h
      {
        prefix:            @prefix,
        suggested_type:    @suggested_type,
        positions:         @positions.map(&:to_h),
        hosts:             @hosts.to_a.sort,
        coverage:          @coverage,
        confidence:        @confidence,
        observation_count: @observation_count,
        sample_values:     @sample_values,
        strategy:          @strategy,
      }
    end

    private

    # Confidence = coverage + linear cross-host boost, capped at 1.0.
    # Single-host proposals get their raw coverage as confidence (no
    # boost). Each additional host adds CROSS_HOST_BOOST_PER_HOST to
    # the score. A proposal supported by ~10 distinct hosts caps out
    # regardless of raw coverage; below that, both signals compose.
    def compute_confidence
      boost = (@hosts.size - 1) * ProposalStrategy::CROSS_HOST_BOOST_PER_HOST
      score = @coverage + boost
      score > 1.0 ? 1.0 : score
    end
  end

  # Pluggable proposal-detection strategies. Each strategy.propose(storage, **opts)
  # returns an array of RecognizerProposal. Adding a new detection rule =
  # add a class with #propose; register it via DEFAULTS.
  module ProposalStrategy
    # Default minimum total matching observations across positions before
    # we'll emit a proposal. Below this the signal is too noisy.
    DEFAULT_MIN_OBSERVATIONS = 20
    # Fraction of sampled observations at affected Positions that must
    # match the proposal pattern.
    DEFAULT_MIN_COVERAGE     = 0.7
    # Minimum number of distinct hosts the proposal must appear at. For
    # single-host corpora this defaults to 1; bumping to 2+ promotes
    # cross-host patterns over host-local ones.
    DEFAULT_MIN_HOSTS        = 1
    # Confidence boost added per additional host beyond the first. A
    # pattern seen on 10+ hosts caps out the boost (+0.45 ≈ 1.0 when
    # combined with any reasonable coverage); single-host patterns get
    # no boost (their coverage IS their confidence).
    CROSS_HOST_BOOST_PER_HOST = 0.05

    # Detects `<prefix>_<alphanumeric>` patterns at slug/opaque_id
    # positions — the GitHub PAT (`ghp_…`), Stripe customer ID (`cus_…`),
    # AWS-style (`sk_test_…` — partial match), Twilio SID-with-letter-
    # prefix family. Restricting the suffix to alphanumeric (no further
    # separators) keeps real slugs (`my-cool-post`, `red_team_member`)
    # from triggering false proposals.
    class PrefixUnderscoreId
      PATTERN = /\A([a-z]+)_([A-Za-z0-9]+)\z/.freeze
      NAME    = :prefix_underscore_id

      def propose(storage,
                  min_observations: DEFAULT_MIN_OBSERVATIONS,
                  min_coverage:     DEFAULT_MIN_COVERAGE,
                  min_hosts:        DEFAULT_MIN_HOSTS)
        per_prefix = Hash.new { |h, k| h[k] = empty_accumulator }

        storage.each_position_stats do |position, stats|
          next unless slug_or_opaque?(stats)

          stats.value_counts.each do |value, count|
            m = PATTERN.match(value) or next
            prefix = "#{m[1]}_"
            acc = per_prefix[prefix]
            acc[:matching_count] += count
            acc[:position_observations] += stats.total unless acc[:positions].include?(position)
            acc[:positions] << position
            acc[:hosts] << position.host
            # Collect every match; we'll sort + cap to a stable top-N at
            # emission time so Ruby and Rust produce identical samples
            # regardless of underlying Hash / map iteration order.
            acc[:matches] << value
          end
        end

        per_prefix.filter_map { |prefix, acc|
          next nil if acc[:matching_count] < min_observations
          next nil if acc[:hosts].size < min_hosts

          coverage = acc[:matching_count].to_f / acc[:position_observations]
          next nil if coverage < min_coverage

          RecognizerProposal.new(
            prefix:            prefix,
            suggested_type:    prefix.chomp("_").to_sym,
            positions:         acc[:positions].to_a,
            hosts:             acc[:hosts],
            coverage:          coverage,
            observation_count: acc[:matching_count],
            # Sort + cap to 5 so Ruby and Rust produce identical samples
            # regardless of underlying Hash / map iteration order. The
            # samples are illustrative for humans; alphabetical is fine.
            sample_values:     acc[:matches].sort.first(5),
            strategy:          NAME,
          )
        }.sort_by { |p| [-p.confidence, p.prefix] }
      end

      private

      def empty_accumulator
        {
          positions:             Set.new,
          hosts:                 Set.new,
          matching_count:        0,
          position_observations: 0,
          matches:               [],
        }
      end

      def slug_or_opaque?(stats)
        dom = stats.type_counts.max_by { |_, c| c }&.first
        dom == :slug || dom == :opaque_id
      end
    end

    DEFAULTS = [PrefixUnderscoreId.new].freeze
  end
end

File: rust/iriq/src/recognizer_proposal.rs
use crate::classifier::SegmentType;
use crate::position::Position;
use crate::position_stats::PositionStats;
use crate::storage::Storage;
use once_cell::sync::Lazy;
use regex::Regex;
use std::collections::{HashMap, HashSet};

#[derive(Debug, Clone)]
pub struct RecognizerProposal {
    pub prefix: String,
    pub suggested_type: String,
    pub positions: Vec<Position>,
    pub hosts: Vec<String>,
    pub coverage: f64,
    pub confidence: f64,
    pub observation_count: usize,
    pub sample_values: Vec<String>,
    pub strategy: String,
}

const CROSS_HOST_BOOST_PER_HOST: f64 = 0.05;

fn compute_confidence(coverage: f64, host_count: usize) -> f64 {
    let score = coverage + CROSS_HOST_BOOST_PER_HOST * (host_count.saturating_sub(1) as f64);
    score.min(1.0)
}

#[derive(Debug, Clone, Copy, Default)]
pub struct ProposalOptions {
    pub min_observations: usize,
    pub min_coverage: f64,
    pub min_hosts: usize,
}

pub const DEFAULT_PROPOSAL_MIN_OBSERVATIONS: usize = 20;
pub const DEFAULT_PROPOSAL_MIN_COVERAGE: f64 = 0.7;
pub const DEFAULT_PROPOSAL_MIN_HOSTS: usize = 1;

fn with_defaults(mut o: ProposalOptions) -> ProposalOptions {
    if o.min_observations == 0 {
        o.min_observations = DEFAULT_PROPOSAL_MIN_OBSERVATIONS;
    }
    if o.min_coverage == 0.0 {
        o.min_coverage = DEFAULT_PROPOSAL_MIN_COVERAGE;
    }
    if o.min_hosts == 0 {
        o.min_hosts = DEFAULT_PROPOSAL_MIN_HOSTS;
    }
    o
}

static PREFIX_UNDERSCORE_ID_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"^([a-z]+)_([A-Za-z0-9]+)$").unwrap());

struct Accumulator {
    positions: HashSet<Position>,
    positions_ordered: Vec<Position>,
    hosts: HashSet<String>,
    matching_count: usize,
    position_observations: usize,
    matches: Vec<String>,
}

pub fn propose_recognizers(
    storage: &dyn Storage,
    opts: ProposalOptions,
) -> Vec<RecognizerProposal> {
    let opts = with_defaults(opts);
    let mut per_prefix: HashMap<String, Accumulator> = HashMap::new();

    let mut visitor = |pos: &Position, stats: &PositionStats| {
        if !slug_or_opaque_dominant(stats) {
            return;
        }
        for (value, count) in &stats.value_counts {
            let Some(caps) = PREFIX_UNDERSCORE_ID_RE.captures(value) else {
                continue;
            };
            let prefix = format!("{}_", caps.get(1).unwrap().as_str());
            let acc = per_prefix.entry(prefix).or_insert_with(|| Accumulator {
                positions: HashSet::new(),
                positions_ordered: Vec::new(),
                hosts: HashSet::new(),
                matching_count: 0,
                position_observations: 0,
                matches: Vec::new(),
            });
            acc.matching_count += count;
            if acc.positions.insert(pos.clone()) {
                acc.positions_ordered.push(pos.clone());
                acc.position_observations += stats.total;
            }
            acc.hosts.insert(pos.host.clone());
            acc.matches.push(value.clone());
        }
    };
    storage.each_position_stats(&mut visitor);

    let mut prefixes: Vec<String> = per_prefix.keys().cloned().collect();
    prefixes.sort();

    let mut out: Vec<RecognizerProposal> = Vec::new();
    for prefix in prefixes {
        let acc = per_prefix.get(&prefix).unwrap();
        if acc.matching_count < opts.min_observations {
            continue;
        }
        if acc.hosts.len() < opts.min_hosts {
            continue;
        }
        let coverage = (acc.matching_count as f64) / (acc.position_observations as f64);
        if coverage < opts.min_coverage {
            continue;
        }
        let mut hosts: Vec<String> = acc.hosts.iter().cloned().collect();
        hosts.sort();
        let mut samples: Vec<String> = acc.matches.clone();
        samples.sort();
        samples.truncate(5);

        let suggested = prefix.trim_end_matches('_').to_string();
        out.push(RecognizerProposal {
            prefix: prefix.clone(),
            suggested_type: suggested,
            positions: acc.positions_ordered.clone(),
            hosts: hosts.clone(),
            coverage,
            confidence: compute_confidence(coverage, hosts.len()),
            observation_count: acc.matching_count,
            sample_values: samples,
            strategy: "prefix_underscore_id".to_string(),
        });
    }

    out.sort_by(|a, b| {
        b.confidence
            .partial_cmp(&a.confidence)
            .unwrap_or(std::cmp::Ordering::Equal)
            .then(a.prefix.cmp(&b.prefix))
    });
    out
}

fn slug_or_opaque_dominant(stats: &PositionStats) -> bool {
    let mut dom = SegmentType::Literal;
    let mut max = 0usize;
    for (&t, &c) in &stats.type_counts {
        if c > max {
            max = c;
            dom = t;
        }
    }
    dom == SegmentType::Slug || dom == SegmentType::OpaqueId
}
